### ISSUE TITLE
test-jetty-osgi test exits/crashes the surefire forked JVM

### ISSUE THREAD

**Issue by joakime** [eclipse/jetty.project#1049]
When building Jetty 9.4 (or master) on Windows, there is a JVM exit/crash reported from surefire.

So far, its happening consistently (every build) on test-jetty-osgi ...

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test) on project test-jetty-osgi: Execution default-test of goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test failed: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
[ERROR] Command was cmd.exe /X /C "C:\jenkins\tools\hudson.model.JDK\jdk8\jre\bin\java -Dmortbay-alpn-boot=C:\jenkins\maven-repositories\0/org/mortbay/jetty/alpn/alpn-boot/8.1.10.v20161026/alpn-boot-8.1.10.v20161026.jar -jar C:\jenkins\workspace\nightlies\jetty-9.4-windows\jetty-osgi\test-jetty-osgi\target\surefire\surefirebooter6317097499128754753.jar C:\jenkins\workspace\nightlies\jetty-9.4-windows\jetty-osgi\test-jetty-osgi\target\surefire\surefire2345530324909305129tmp C:\jenkins\workspace\nightlies\jetty-9.4-windows\jetty-osgi\test-jetty-osgi\target\surefire\surefire_651711783400904617222tmp"
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :test-jetty-osgi
```

**Comment by janbartel** [comment]
Doesn't only happen on windows, happens sometimes on linux.

I can never reproduce this locally, without inserting an explicit System.exit in a test case.

Reading the Surefire plugin faq: http://maven.apache.org/surefire/maven-surefire-plugin/faq.html#vm-termination
it seems to me like a CI environment problem: is there any chance the box is running out of resources and killing the forked process?

**Comment by joakime** [comment]
Here's a capture of a crashed VM (on Windows)

```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.eclipse.jetty.osgi.test.TestJettyOSGiBootContextAsService
14 [main] INFO org.ops4j.pax.exam.spi.DefaultExamSystem - Pax Exam System (Version: 4.5.0) created.
54 [main] INFO org.ops4j.pax.exam.junit.impl.ProbeRunner - creating PaxExam runner for class org.eclipse.jetty.osgi.test.TestJettyOSGiBootContextAsService
DEFAULT_HTTP_PORT=61062
DEFAULT_SSL_PORT=61063
229 [main] INFO org.ops4j.pax.exam.junit.impl.ProbeRunner - running test class org.eclipse.jetty.osgi.test.TestJettyOSGiBootContextAsService
298 [main] INFO org.ops4j.exec.DefaultJavaRunner - DefaultJavaRunner completed successfully

org.eclipse.jetty.util[org.eclipse.jetty.osgi.boot.internal.serverfactory.ServerInstanceWrapper] : File does not exist file://C/code/jetty/jetty.project/jetty-osgi/test-jetty-osgi/src/test/config/etc/jetty.xml Ignored FQCN: org.eclipse.jetty.util.log.Slf4jLog
org.eclipse.jetty.util[org.eclipse.jetty.osgi.boot.internal.serverfactory.ServerInstanceWrapper] : Configuration error in file://C:/code/jetty/jetty.project/jetty-osgi/test-jetty-osgi/src/test/config/etc/jetty.xml Ignored FQCN: org.eclipse.jetty.util.log.Slf4jLog
org.eclipse.jetty.util[org.eclipse.jetty.osgi.boot.internal.serverfactory.DefaultJettyAtJettyHomeHelper] :  Ignored FQCN: org.eclipse.jetty.util.log.Slf4jLog
java.lang.IllegalStateException: No such jetty server config file: file://C/code/jetty/jetty.project/jetty-osgi/test-jetty-osgi/src/test/config/etc/jetty.xml
        at org.eclipse.jetty.osgi.boot.internal.serverfactory.ServerInstanceWrapper.configure(ServerInstanceWrapper.java:153)
        at org.eclipse.jetty.osgi.boot.internal.serverfactory.DefaultJettyAtJettyHomeHelper.startJettyAtJettyHome(DefaultJettyAtJettyHomeHelper.java:179)
        at org.eclipse.jetty.osgi.boot.JettyBootstrapActivator.start(JettyBootstrapActivator.java:92)
        at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:771)
        at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:1)
        at java.security.AccessController.doPrivileged(Native Method)
        at org.eclipse.osgi.internal.framework.BundleContextImpl.startActivator(BundleContextImpl.java:764)
        at org.eclipse.osgi.internal.framework.BundleContextImpl.start(BundleContextImpl.java:721)
        at org.eclipse.osgi.internal.framework.EquinoxBundle.startWorker0(EquinoxBundle.java:936)
        at org.eclipse.osgi.internal.framework.EquinoxBundle$EquinoxModule.startWorker(EquinoxBundle.java:319)
        at org.eclipse.osgi.container.Module.doStart(Module.java:571)
        at org.eclipse.osgi.container.Module.start(Module.java:439)
        at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1582)
        at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1562)
        at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1533)
        at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1476)
        at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)
        at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
        at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:340)
java.lang.IllegalStateException: No such jetty server config file: file://C/code/jetty/jetty.project/jetty-osgi/test-jetty-osgi/src/test/config/etc/jetty.xml
        at org.eclipse.jetty.osgi.boot.internal.serverfactory.ServerInstanceWrapper.configure(ServerInstanceWrapper.java:153)
        at org.eclipse.jetty.osgi.boot.internal.serverfactory.DefaultJettyAtJettyHomeHelper.startJettyAtJettyHome(DefaultJettyAtJettyHomeHelper.java:179)
        at org.eclipse.jetty.osgi.boot.JettyBootstrapActivator.start(JettyBootstrapActivator.java:92)
        at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:771)
        at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:1)
        at java.security.AccessController.doPrivileged(Native Method)
        at org.eclipse.osgi.internal.framework.BundleContextImpl.startActivator(BundleContextImpl.java:764)
        at org.eclipse.osgi.internal.framework.BundleContextImpl.start(BundleContextImpl.java:721)
        at org.eclipse.osgi.internal.framework.EquinoxBundle.startWorker0(EquinoxBundle.java:936)
        at org.eclipse.osgi.internal.framework.EquinoxBundle$EquinoxModule.startWorker(EquinoxBundle.java:319)
        at org.eclipse.osgi.container.Module.doStart(Module.java:571)
        at org.eclipse.osgi.container.Module.start(Module.java:439)
        at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1582)
        at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1562)
        at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1533)
        at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1476)
        at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)
        at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
        at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:340)
DEFAULT_HTTP_PORT=61062
DEFAULT_SSL_PORT=61063

10560 [main] INFO org.ops4j.exec.DefaultJavaRunner - Platform has been shutdown.
10566 [main] INFO org.ops4j.pax.exam.spi.reactors.ReactorManager - suite finished
Tests run: 2, Failures: 0, Errors: 1, Skipped: 1, Time elapsed: 10.58 sec <<< FAILURE! - in org.eclipse.jetty.osgi.test.TestJettyOSGiBootContextAsService
testContextHandlerAsOSGiService(org.eclipse.jetty.osgi.test.TestJettyOSGiBootContextAsService)  Time elapsed: 10.335 sec  <<< ERROR!
java.net.ConnectException: Connection refused: no further information

Running org.eclipse.jetty.osgi.test.TestJettyOSGiBootCore
10566 [main] INFO org.ops4j.pax.exam.junit.impl.ProbeRunner - creating PaxExam runner for class org.eclipse.jetty.osgi.test.TestJettyOSGiBootCore
10586 [main] INFO org.ops4j.pax.exam.junit.impl.ProbeRunner - running test class org.eclipse.jetty.osgi.test.TestJettyOSGiBootCore
10590 [main] INFO org.ops4j.exec.DefaultJavaRunner - DefaultJavaRunner completed successfully

DEFAULT_HTTP_PORT=61062
DEFAULT_SSL_PORT=61063

15871 [main] INFO org.ops4j.exec.DefaultJavaRunner - Platform has been shutdown.
15872 [main] INFO org.ops4j.pax.exam.spi.reactors.ReactorManager - suite finished
Tests run: 2, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 5.272 sec - in org.eclipse.jetty.osgi.test.TestJettyOSGiBootCore
Running org.eclipse.jetty.osgi.test.TestJettyOSGiBootHTTP2
15873 [main] INFO org.ops4j.pax.exam.junit.impl.ProbeRunner - creating PaxExam runner for class org.eclipse.jetty.osgi.test.TestJettyOSGiBootHTTP2
15895 [main] INFO org.ops4j.pax.exam.junit.impl.ProbeRunner - running test class org.eclipse.jetty.osgi.test.TestJettyOSGiBootHTTP2
15899 [main] INFO org.ops4j.exec.DefaultJavaRunner - DefaultJavaRunner completed successfully


Results :

Tests in error:
  TestJettyOSGiBootContextAsService.testContextHandlerAsOSGiService » Connect Co...

Tests run: 4, Failures: 0, Errors: 1, Skipped: 2

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] Jetty :: OSGi ...................................... SUCCESS [  1.881 s]
[INFO] Jetty :: OSGi :: Boot .............................. SUCCESS [  3.922 s]
[INFO] Jetty :: OSGi :: Boot JSP .......................... SUCCESS [  0.944 s]
[INFO] Jetty :: OSGi :: Boot :: Warurl .................... SUCCESS [  0.608 s]
[INFO] Jetty :: OSGi :: HttpService ....................... SUCCESS [  0.467 s]
[INFO] Jetty :: OSGi :: Test WebApp ....................... SUCCESS [  0.425 s]
[INFO] Jetty :: OSGi :: Test Context ...................... SUCCESS [  0.408 s]
[INFO] Jetty :: OSGi :: WebApp Fragment ................... SUCCESS [  0.289 s]
[INFO] Jetty :: OSGi ALPN Fragment ........................ SUCCESS [  0.222 s]
[INFO] Jetty :: OSGi :: Test .............................. FAILURE [ 23.769 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 33.759 s
[INFO] Finished at: 2017-03-09T07:02:44-07:00
[INFO] Final Memory: 82M/1033M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test) on project test-jetty-osgi: Execution default-test of goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test failed: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
[ERROR] Command was cmd.exe /X /C ""C:\Program Files\Java\jdk1.8.0_121\jre\bin\java" -Dmortbay-alpn-boot=C:\Users\joakim\.m2\repository/org/mortbay/jetty/alpn/alpn-boot/8.1.11.v20170118/alpn-boot-8.1.11.v20170118.jar -jar C:\code\jetty\jetty.project\jetty-osgi\test-jetty-osgi\target\surefire\surefirebooter4238790219302624456.jar C:\code\jetty\jetty.project\jetty-osgi\test-jetty-osgi\target\surefire\surefire8359516226714691871tmp C:\code\jetty\jetty.project\jetty-osgi\test-jetty-osgi\target\surefire\surefire_01595307849153515365tmp"
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :test-jetty-osgi
```

Note: I upgraded PAXEXAM to 4.5.0 on my local machine to get around a Windows Bug w ith PAXEXAM - https://ops4j1.jira.com/browse/PAXEXAM-701

I'm retesting with PAXEXAM 4.9.1 now...

**Comment by joakime** [comment]
I've noticed that if the maven JVM crashes, the paxexam spawned JVM is still running.

Which on windows is really really bad, as now the directories in the `test-jetty-osgi/target/` are now file-locked by that JVM and the next maven build will always fail.

**Comment by joakime** [comment]
Here's the command line of the zombie JVM ...

```
VM Arguments:
jvm_args: 
  -Xbootclasspath/p:C:\Users\joakim\.m2\repository/org/mortbay/jetty/alpn/alpn-boot/8.1.11.v20170118/alpn-boot-8.1.11.v20170118.jar 
  -Dpax.swissbox.framework.rmi.port=21000 
  -Dpax.exam.logging=none 
  -Djava.protocol.handler.pkgs=org.ops4j.pax.url 
  -Dorg.ops4j.pax.logging.DefaultServiceLog.level=WARN 
  -Dpax.swissbox.framework.rmi.name=ExamRemoteFramework-a7bc9680-0bd2-4e17-bcaa-56400106f596 
  -Djetty.http.port=61280 
  -Dpax.exam.invoker=junit 
  -Dpax.exam.service.timeout=10000 
  -Dorg.eclipse.jetty.osgi.LEVEL=WARN 
  -Djetty.base=C:\code\jetty\jetty.project\jetty-osgi\test-jetty-osgi\src\test\config 
  -Djetty.ssl.port=61281 
  -Djetty.etc.config.urls=file://C:\code\jetty\jetty.project\jetty-osgi\test-jetty-osgi\src\test\config\etc/jetty.xml
    ;file://C:\code\jetty\jetty.project\jetty-osgi\test-jetty-osgi\src\test\config\etc/jetty-ssl.xml
    ;file://C:\code\jetty\jetty.project\jetty-osgi\test-jetty-osgi\src\test\config\etc/jetty-https.xml
    ;
    ;file://C:\code\jetty\jetty.project\jetty-osgi\test-jetty-osgi\src\test\config\etc/jetty-http2.xml
    ;file://C:\code\jetty\jetty.project\jetty-osgi\test-jetty-osgi\src\test\config\etc/jetty-deployer.xml
    ;file://C:\code\jetty\jetty.project\jetty-osgi\test-jetty-osgi\src\test\config\etc/jetty-testrealm.xml
  -Djetty.home=C:\code\jetty\jetty.project\jetty-osgi\test-jetty-osgi\src\test\config 
java_command: 
  org.ops4j.pax.swissbox.framework.RemoteFrameworkImpl 
  -Forg.osgi.framework.bootdelegation=sun.* 
  -Forg.osgi.framework.storage=C:\code\jetty\jetty.project\jetty-osgi\test-jetty-osgi\target\1489069070403-0
java_class_path (initial): 
    /C:/Users/joakim/.m2/repository/org/eclipse/osgi/3.10.0-v20140606-1445/osgi-3.10.0-v20140606-1445.jar
    ;/C:/Users/joakim/.m2/repository/org/ops4j/pax/swissbox/pax-swissbox-framework/1.8.2/pax-swissbox-framework-1.8.2.jar
    ;/C:/Users/joakim/.m2/repository/org/ops4j/pax/swissbox/pax-swissbox-tracker/1.8.2/pax-swissbox-tracker-1.8.2.jar
Launcher Type: SUN_STANDARD
```

Maven error at this time:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test) on project test-jetty-osgi: Execution default-test of goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test failed: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
[ERROR] Command was cmd.exe /X /C ""C:\Program Files\Java\jdk1.8.0_121\jre\bin\java" -Dmortbay-alpn-boot=C:\Users\joakim\.m2\repository/org/mortbay/jetty/alpn/alpn-boot/8.1.11.v20170118/alpn-boot-8.1.11.v20170118.jar -jar C:\code\jetty\jetty.project\jetty-osgi\test-jetty-osgi\target\surefire\surefirebooter3751283670025095682.jar C:\code\jetty\jetty.project\jetty-osgi\test-jetty-osgi\target\surefire\surefire2160317377968171327tmp C:\code\jetty\jetty.project\jetty-osgi\test-jetty-osgi\target\surefire\surefire_06504184805908749099tmp"
```

**Comment by joakime** [comment]
Using PAXEXAM 4.9.1 makes no difference here.
It does, however, fix the [PAXEXAM-701](https://ops4j1.jira.com/browse/PAXEXAM-701) issue.

**Comment by joakime** [comment]
Looks like the JVM crash can be gotten around with a downgrade of surefire to the 2.18.x series.
There's a bunch of reported bugs against the surefire/failsafe 2.19.x maven plugin series when using OSGi.

**Comment by janbartel** [comment]
Some of those urls look dodgy. I'll take a look at the code.

**Comment by janbartel** [comment]
So I've fixed the dodgy urls. Can this issue be closed now, and only work on #1387?

**Comment by joakime** [comment]
Not crashing anymore, considering this closed.
